src/protocol/src/lib.rs

```diff
@@ -77,6 +83,10 @@ pub const VERSION_GSSENC: Version = (1234 << 16) + 5680;
 /// `Error` type in protocol `Result`. Indicates that something went not well
 #[derive(Debug, PartialEq)]
 pub enum Error {
+    /// Notifies that the specified connection ID needs to be cancelled
+    CancelRequest(i32),
```

Comment:
I think query cancelation is not a error flow for PG wire protocol (if it so can you point me into PostgreSQL doc about it) so then `CancenlRequest(i32)` and `Ok((Receiver, Sender))` has to be extracted into separate enum.
Can you introduce enum in `protocol` module like:
```rust
pub enum ClientRequest {
   Connection(impl Receiver, impl Sender),
   QueryCancelation(i32), 
}
```
BTW, can you also use one of added type alias `ConnId` or `ConnSecretKey` otherwise it is not clear which id should be used for cancelation.

Then `protocol::hand_shake()` will have return type `io::Result<Result<ClientRequest>>`.

Also I think `protocol::hand_shake` should be renamed into `protocol::accept_client_request`